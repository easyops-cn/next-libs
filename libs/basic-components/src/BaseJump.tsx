import React from "react";
import { Input, Icon } from "antd";
import { withTranslation, WithTranslation } from "react-i18next";
import { Link } from "./Link";
import style from "./BaseJump.module.css";
import { Subject } from "rxjs/internal/Subject";
import { debounceTime, distinctUntilChanged } from "rxjs/operators";
import { getInstanceShowName } from "@libs/cmdb-utils";
interface BaseJumpProps extends WithTranslation {
  isBackendSearch?: boolean;
  instanceList?: [];
  modelDefine?: {};

  fetchData: any;
  title: string;
  nameKey: string;
}
interface BaseJumpState {
  _show: boolean;
  showList: any[];
  inputVal: string;
  cacheList: any[];
  instanceName?: string;
}
interface InstanceItem {
  instanceId: string;
  name?: string;
  [key: string]: string;
}
export class LegacyBaseJump extends React.Component<
  BaseJumpProps,
  BaseJumpState
> {
  searchTerm$: Subject<string>;
  wrapperRef: any;
  constructor(props: any, context: any) {
    super(props, context);
    this.setWrapperRef = this.setWrapperRef.bind(this);
    this.handleClickOutside = this.handleClickOutside.bind(this);
    this.searchTerm$ = new Subject<string>();
    this.state = {
      _show: false,
      showList: [],
      inputVal: "",
      cacheList: [] //缓存一开始空值搜索的列表
    };
  }
  renderList(list: InstanceItem[] = [], nameKey: string) {
    const showList = list.map(item => {
      return (
        <li key={item.instanceId}>
          <Link to={item.instanceId}>
            {getInstanceShowName(item, [nameKey])}
          </Link>
        </li>
      );
    });
    return showList;
  }
  async componentDidMount() {
    // @ts-ignore
    document.addEventListener("click", this.handleClickOutside);
    await this.searchTerm$
      .pipe(
        debounceTime(500),
        distinctUntilChanged()
      )
      .subscribe(res => {
        this.fetchInstanceList(res);
      });
  }
  componentWillUnmount() {
    // @ts-ignore
    document.removeEventListener("click", this.handleClickOutside);
  }
  async fetchInstanceList(query: string) {
    const q = (query || "").toLowerCase();
    if (q === "" || q == null) {
      this.setState({
        showList: this.state.cacheList
      });
      return;
    }
    const list = await this.props.fetchData(q);
    this.setState({
      showList: list
    });
  }
  async toggleElementState() {
    this.setState(prevState => ({
      _show: !prevState._show
    }));
    if (!this.state.cacheList.length) {
      const cacheList = await this.props.fetchData("");
      this.setState({
        showList: cacheList,
        cacheList
      });
    }
  }
  setWrapperRef(node: any) {
    this.wrapperRef = node;
  }
  handleClickOutside(event: React.ChangeEvent<HTMLInputElement>) {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({
        _show: false
      });
    }
    return;
  }
  handleSearchInstance(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({
      inputVal: event.target.value
    });
    this.searchTerm$.next(event.target.value);
  }
  render(): React.ReactNode {
    const { _show, showList, inputVal, instanceName } = this.state;
    const { nameKey } = this.props;

    return (
      <div style={{ position: "relative" }} ref={this.setWrapperRef}>
        <div
          onClick={this.toggleElementState.bind(this)}
          className={style.customTitleContainer}
        >
          <span className={style.customTitle}>{this.props.title}</span>
          <span
            style={{ marginLeft: "5px", marginTop: "5px", fontSize: "4px" }}
          >
            <Icon type={_show ? "caret-up" : "caret-down"} />
          </span>
        </div>
        <div className={(_show ? "" : style.hidden) + " " + style.gridCard}>
          <div className={style.searchBoxHeading}>
            <div className={style.searchBar}>
              <Input
                value={inputVal}
                onChange={this.handleSearchInstance.bind(this)}
                suffix={<Icon type="search" />}
              />
            </div>
          </div>
          <div className={style.searchContainer}>
            {showList.length > 0 ? (
              <ul className={style.searchList}>
                {this.renderList(showList, this.props.nameKey)}
              </ul>
            ) : (
              <span className={style.zeroSearchResult}>查询结果为空</span>
            )}
          </div>
        </div>
      </div>
    );
  }
}

export const BaseJump = withTranslation("BaseJump")(LegacyBaseJump);
