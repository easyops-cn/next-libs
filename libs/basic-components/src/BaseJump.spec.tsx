import React from "react";
import { shallow } from "enzyme";
import { BaseJump, LegacyBaseJump } from "./BaseJump";
import { InstanceApi } from "@sdk/cmdb-sdk";

jest.mock("@sdk/cmdb-sdk");
describe("BrickInstanceJump", () => {
  let wrapper: any;
  let component: LegacyBaseJump;
  beforeAll(() => {
    wrapper = shallow(
      <BaseJump nameKey="name" title="xxx" fetchData={() => {}} />
    );
    component = wrapper.instance() as LegacyBaseJump;
  });
  it("should work", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("setWrapperRef should work", () => {
    component.setWrapperRef("xxx");
    expect(component.wrapperRef).toEqual("xxx");
  });
  it("handleClickOutside", () => {
    const spyOnSetState = jest.spyOn(component, "setState");
    component.wrapperRef = {};
    component.wrapperRef.contains = () => false;
    component.handleClickOutside({ target: "xxx" } as any);
    expect(spyOnSetState).toBeCalledWith({
      _show: false
    });
  });
  it("toggleElementState should work", () => {
    const spyOnSetState = jest.spyOn(component, "setState");
    component.toggleElementState();
    expect(spyOnSetState).toBeCalled();
  });
  it("handleSearchInstance should work", async () => {
    jest
      .spyOn(InstanceApi, "postSearch")
      // eslint-disable-next-line
      .mockResolvedValue({ list: [], total: 0, page: 1, page_size: 10 });

    await component.componentDidMount();
    const spyOnBackQueryInstanceList = jest.spyOn(
      component,
      "fetchInstanceList"
    );
    component.handleSearchInstance({ target: { value: "xxx" } } as any);
    await jest.runAllTimers();
    expect(spyOnBackQueryInstanceList).toBeCalledWith("xxx");
  });
});
