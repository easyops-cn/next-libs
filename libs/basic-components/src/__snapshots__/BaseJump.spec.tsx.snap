// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickInstanceJump should work 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="customTitleContainer"
    onClick={[Function]}
  >
    <span
      className="customTitle"
    >
      xxx
    </span>
    <span
      style={
        Object {
          "fontSize": "4px",
          "marginLeft": "5px",
          "marginTop": "5px",
        }
      }
    >
      <Icon
        type="caret-down"
      />
    </span>
  </div>
  <div
    className="hidden gridCard"
  >
    <div
      className="searchBoxHeading"
    >
      <div
        className="searchBar"
      >
        <Input
          onChange={[Function]}
          suffix={
            <Icon
              type="search"
            />
          }
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="searchContainer"
    >
      <span
        className="zeroSearchResult"
      >
        查询结果为空
      </span>
    </div>
  </div>
</div>
`;
